service/frontend/workflow_handler.go

```diff
@@ -4273,30 +4284,42 @@ func (wh *WorkflowHandler) getHistoryReverse(
 	nextPageToken []byte,
 	branchToken []byte,
 ) (*historypb.History, []byte, int64, error) {
+
+	// readFullPageEventsReverse reads a full page of history events from history service in reverse orcer. Due to
+	// storage format of V2 History it is not guaranteed that pageSize amount of data is returned. Function returns the
+	// list of history events, the size of data read, the next page token, and an error if present.
+	readFullPageEventsReverse := func(
+		req *historyservice.ReadHistoryBranchReverseRequest,
+	) ([]*historypb.HistoryEvent, int, []byte, error) {
+		var historyEvents []*historypb.HistoryEvent
+		size := 0
+		for {
+			response, err := wh.historyClient.ReadHistoryBranchReverse(ctx, req)
```

Comment:
We _must_ truncate the number of historyEvents that we return to the requested page size, or we may return up to `pageSize-1` extra elements. For example, what happens if we read `pageSize-1` elements, and then `pageSize` elements? 

We _should_ fix this by decrementing `req.PageSize` by `len(historyEvents)` at the end of each iteration because that will fix the problem, and it will likely lead to performance gains.

I'd also like us to extract this to a common method with its own dedicated test for the example above. After that is done, we should reuse it in the method above. Or, better yet, we should just fix it at the history client level, but I think that's too much work for this PR.

Reply:
bingo! this is what I was eluding to in my previous responses. But this PR is about introducing the gRPC call, but about fixing any algorithmic issues. For now, treat this as a verbatim port over from `history_manager_util.go` and I'm neither making any logic more correct or worse.